### DISCUSSION TITLE
Unclear on what Copilot sends to servers

### DISCUSSION THREAD

**Discussion by B45man** [community/community#13846] [Copilot Conversations] [answered]
I'm working for a large company and got invited to the Copilot beta which I'm happy about. However, there are questions within the company on how safe Copilot is to use. 

I'm reading the terms of service and telemetry page, but it's not very clear on what actually gets sent to external servers from my code. I'm a .Net web developer, so as an example asume I'm working in C# with .cs files. 

Based on how I'm reading the terms of service, these things I understand will for sure, 100% get sent to Microsoft:

- The snippet I've typed in, such as "generate a unit test that tests if method X throws an exception if input is null"
- The unit test that is generated from this snippet.
- Any changes I make to this generated unit test.

What "may" get sent to Microsoft:
- The complete .cs file that I typed the snippet into
- The URL to my Azure Devops git repository
- Any files containing code relevant to the snippet. So in my unit test snippet above, if method X is in another class in another .cs file, that entire .cs file is sent to Microsoft as well. If method X also references code in other files, those files are sent over too. 

Is my understanding correct?

**Comment by davecheney** [comment, marked as answer]
Yes

**Reply by B45man** [reply]
Thanks for your time. As an additional question which is not 100% clear to me, do the .cs files in my example get permanently stored by Microsoft as well, or are those just used for analysis and then discarded?

**Reply by davecheney** [reply]
I’m sure you’ve already read it, but the answers I’m quoting from are here https://docs.github.com/en/github/copilot/about-github-copilot-telemetry. The retention duration is not defined.

**Reply by danmaxis** [reply]
But can we consider it "Safe for work"?

**Reply by EliasBorrajo** [reply]
This is a huge concern for mee too, i didn't enabled copilot on other IDE i'm working with, by fear any sensitive data would be sent.

So i have a playground IDE for my personnal projects, where i play with copilot to try it out and see how it works out.

But this has to be clarified, because this would be a red flag, a big "no" to use this intersting tool.

**Reply by milvone** [reply]
What about non-code files like text files? Or DB dump SQL files? Is there any limitations on what content / filetypes will be sent or can we summarize it "what ever is open in your editor plus anything that is considered a link (module includes etc.)"?

**Reply by shadow-cs** [reply]
@davecheney  Followup question: is it somehow configurable (which files get sent as context as part of the prompt)?

I'm aware of https://docs.github.com/en/copilot/configuring-github-copilot/configuring-github-copilot-in-your-environment, but how does VSC plugin itself work internally? If I disable the file type (i.e. `"plaintext": false`) will it disable sending `plaintext` files as part of the prompt?

Is there another way of controlling what gets sent?

**Reply by davecheney** [reply]
> Followup question: is it somehow configurable (which files get sent as context as part of the prompt)?

not at the moment.

**Reply by shadow-cs** [reply]
> not at the moment.

@davecheney meaning any file in the workspace (regardless of the setting inside `github.copilot.enable`) may be sent? This includes files we'd place in `.gitignore` etc. (which may contain secrets).

**Reply by davecheney** [reply]
at the moment, yes.

**Reply by shadow-cs** [reply]
Thanks for the clarification, I hope we'll get some more configuration at our disposal. The [`.copilotignore` suggestion](https://github.com/orgs/community/discussions/10305) looks quite nice.

**Reply by davecheney** [reply]
You're welcome. I wasn't sure if copilotignore was available yet.

**Reply by shadow-cs** [reply]
@davecheney so copilotignore is on the roadmap? Or some other method to tell Copilot, don't look in this file perhaps?
